### PR TITLE
micro-fix(security): mitigate ast.Pow DoS and enforce safe_eval timeout

### PR DESCRIPTION
**Pull request by sundaram2021**
## Description

Mitigates denial-of-service vectors in `framework.graph.safe_eval` by hardening power evaluation and enforcing a bounded execution timeout for edge-condition expressions.

The change is intentionally scoped to `safe_eval` so existing graph execution behavior stays intact outside of rejecting pathological or abusive expressions. Existing callers do not need to change: `safe_eval(expr, context)` remains valid.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)

## Changes Made

- Replace direct `ast.Pow` evaluation with a bounded helper that rejects oversized exponents
- Reject integer power expressions whose estimated result size would exceed safe limits
- Add a default `safe_eval` execution timeout of `100ms` for edge-condition evaluation
- Keep timeout configurable via `timeout_ms`, including `timeout_ms=None` to disable it explicitly
- Preserve backward compatibility for existing `safe_eval(expr, context)` call sites
- Add regression coverage for power-operator DoS cases and timeout behavior

## Testing

Describe the tests you ran to verify your changes:

- [x] Unit tests pass (`cd core && pytest tests/`)
- [x] Lint passes (`cd core && ruff check .`)
- [x] Manual testing performed

Manual testing:
- Ran `cd core && uv run pytest tests/ -v`
- Result: `1562 passed, 43 skipped`
- Ran `cd core && uv run ruff check .`
- Ran `cd core && uv run ruff format --check .`
- Ran focused `safe_eval` and conditional-edge tests during development
- Benchmarked representative edge-condition expressions locally; typical evaluations were ~`0.01ms` to `0.05ms`, which supports a `100ms` default timeout with ample headroom for legitimate use

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

* **New Features**
  * Added configurable execution timeout for evaluations (default 100 ms) to prevent long-running runs; pass timeout_ms=None to disable.

* **Improvements**
  * Clearer validation and errors for timeout values (rejects non-positive timeouts).
  * Strengthened exponentiation safety: rejects excessively large exponents/results with explicit errors.

### CODE REVIEW

File: core/framework/graph/safe_eval.py [hunk 1]
```diff
@@ -1,7 +1,76 @@
...
+        raise TimeoutError(_timeout_message(timeout_ms))
+
+    old_handler = signal.getsignal(signal.SIGALRM)
+    signal.signal(signal.SIGALRM, _handle_timeout)
+    signal.setitimer(signal.ITIMER_REAL, timeout_ms / 1000)
+    try:
+        yield
+    finally:
+        signal.setitimer(signal.ITIMER_REAL, 0)
+        signal.signal(signal.SIGALRM, old_handler)
```
[the remark this answers is not in the document]
Comment (sundaram2021) [reply core/framework/graph/safe_eval.py:78]: fixed

File: core/framework/graph/safe_eval.py [hunk 2]
```diff
@@ -1,7 +1,84 @@
...
+    if isinstance(exp, (int, float)) and abs(exp) > MAX_POWER_ABS_EXPONENT:
+        raise ValueError(f"Power exponent exceeds safe limit ({MAX_POWER_ABS_EXPONENT})")
+
+    if isinstance(base, int) and isinstance(exp, int) and exp > 0:
+        abs_base = abs(base)
+        if abs_base > 1:
+            # Estimate bit growth instead of materializing a huge integer.
+            estimated_bits = exp * (abs_base.bit_length() - 1) + 1
+            if estimated_bits > MAX_POWER_RESULT_BITS:
+                raise ValueError("Power operation exceeds safe size limit")
```
[the remark this answers is not in the document]
Comment (sundaram2021) [reply core/framework/graph/safe_eval.py:28]: done
